### PR TITLE
Some Bloodmoon Change

### PR DESCRIPTION
**Pull request by Ultradragon005**

### CODE REVIEW

**Comment by Ultradragon005** [comment]
ported colelction extensiosn from EHR

**Comment by Tommy-XL** [comment]
Lol

image

**Comment by Ultradragon005** [comment]
bruh what the heck

File: Modules/Extensions/CollectionExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            int n = list.Count;
+            var r = IRandom.Instance;
+            while (n > 1)
+            {
+                n--;
+                int k = r.Next(n + 1);
+                (list[n], list[k]) = (list[k], list[n]);
+            }
+            return list;
+        }
```
Comment (Tommy-XL) [inline Modules/Extensions/CollectionExtensions.cs:41]: This already exists in the code

image

**Comment by Ultradragon005** [comment]
fixed github tripping ballz

**Comment by Gurge44** [comment]
I would've liked to see a link to my original file at the top, guess they don't do that here

**Comment by Ultradragon005** [comment]
oh okay, gonna ask tommy to add this so I don't need to make another pr

**Comment by Tommy-XL** [comment]
> I would've liked to see a link to my original file at the top, guess they don't do that here

Ok, i will added it

**Comment by Tommy-XL** [comment]
> I would've liked to see a link to my original file at the top, guess they don't do that here

https://github.com/0xDrMoe/TownofHost-Enhanced/commit/08b1dcf0005bcc53f32710041d41f10eceab8b8e

**Comment by Ultradragon005** [comment]
> Also worth mentioning that this `Do` method will make the iteration run in parallel between the elements, so the order of completion may be different from the order of elements in the collection, but it makes a huge performance improvement (and in 99% of cases we don't care about the order of completion) But this may not work well with BepInEx (like async)

Multi-Threading at it's finest 😼 but unfortunately it can lead to crash sometimes

**Comment by Ultradragon005** [comment]
but ig the other parrarel methods run just fine, I noticed there was no lag when opening the settings manager now.

**Comment by Gurge44** [comment]
> but ig the other parrarel methods run just fine, I noticed there was no lag when opening the settings manager now.

that was because of another thing ported from EHR lol

**Comment by Gurge44** [comment]
> > Also worth mentioning that this `Do` method will make the iteration run in parallel between the elements, so the order of completion may be different from the order of elements in the collection, but it makes a huge performance improvement (and in 99% of cases we don't care about the order of completion) But this may not work well with BepInEx (like async)
> 
> Multi-Threading at it's finest 😼 but unfortunately it can lead to crash sometimes

BepInEx isn't really friends with async or parallel lol

**Comment by Ultradragon005** [comment]
> > but ig the other parrarel methods run just fine, I noticed there was no lag when opening the settings manager now.
> 
> that was because of another thing ported from EHR lol

oh.

**Comment by Gurge44** [comment]
Basically everything in System.Threading is useless

**Comment by Gurge44** [comment]
So I think you should remove the other Parallel stuff because it will crash the game too

**Comment by Ultradragon005** [comment]
ye I realize, gonna do it in my lates pr

**Comment by Tommy-XL** [comment]
Game crash due `CustomRpcSender`  so I temporarily removed the code

image

**Pull request merged**